src/track/serato/markers.cpp

```diff
@@ -0,0 +1,229 @@
+//     Serato format | 00 00 01 4c  00000000000000000000000101001100
+//                   |
+//     3-byte RGB    |    cc 88 00       110 0110010 0010000 0000000
+//     Serato format | 06 32 10 00  00000110001100100001000000000000
+//
+// See this for details:
+// https://github.com/Holzhaus/serato-tags/blob/master/docs/serato_markers_.md#color-format
+
+QRgb colorToRgb(quint8 w, quint8 x, quint8 y, quint8 z) {
+    quint8 b = (z & 0x7F) | ((y & 0x01) << 7);
+    quint8 g = ((y & 0x7F) >> 1) | ((x & 0x03) << 6);
+    quint8 r = ((x & 0x7F) >> 2) | ((w & 0x07) << 5);
+    return QRgb(r << 16) | (g << 8) | b;
+}
+
+QRgb colorToRgb(quint32 color) {
+    return colorToRgb(
+            (color >> 24) & 0xFF,
+            (color >> 16) & 0xFF,
+            (color >> 8) & 0xFF,
+            color & 0xFF);
+}
+
+quint32 colorFromRgb(quint8 r, quint8 g, quint8 b) {
+    quint8 z = b & 0x7F;
+    quint8 y = ((b >> 7) | (g << 1)) & 0x7F;
+    quint8 x = ((g >> 6) | (r << 2)) & 0x7F;
+    quint8 w = (r >> 5);
+    return (w << 24) | (x << 16) | (y << 8) | z;
+}
+
+quint32 colorFromRgb(QRgb rgb) {
+    return colorFromRgb(qRed(rgb), qGreen(rgb), qBlue(rgb));
+}
+
+}
+
+namespace mixxx {
+
+QByteArray SeratoMarkersEntry::data() const {
+    QByteArray data;
+    data.resize(kEntrySize);
+
+    QDataStream stream(&data, QIODevice::WriteOnly);
+    stream.setVersion(QDataStream::Qt_5_0);
+    stream.setByteOrder(QDataStream::BigEndian);
+    stream << (quint8)((m_startPosition == -1) ? 0x7F : 0x00)
+           << (quint32)((m_startPosition == -1) ? 0x7F7F7F7F : m_startPosition)
+           << (quint8)((m_endPosition == -1) ? 0x7F : 0x00)
+           << (quint32)((m_endPosition == -1) ? 0x7F7F7F7F : m_endPosition);
+    stream.writeRawData("\x00\x7F\x7F\x7F\x7F\x7F", 6);
+    stream << (quint32)colorFromRgb(m_color)
+           << (quint8)m_type
+           << (quint8)m_isLocked;
+    return data;
+}
+
+SeratoMarkersEntryPointer SeratoMarkersEntry::parse(const QByteArray& data) {
+    if (data.length() != kEntrySize) {
+        qWarning() << "Parsing SeratoMarkersEntry failed:"
+                   << "Length" << data.length() << "!=" << kEntrySize;
+        return nullptr;
+    }
+
+    quint8 type;
+    quint8 startPositionSet;
+    quint8 endPositionSet;
+    quint32 startPositionRaw;
+    quint32 endPositionRaw;
+    quint32 colorRaw;
+    bool isLocked;
+    char buffer[6];
+
+    QDataStream stream(data);
+    stream.setVersion(QDataStream::Qt_5_0);
+    stream.setByteOrder(QDataStream::BigEndian);
+    stream >> startPositionSet >> startPositionRaw >> endPositionSet >> endPositionRaw;
+
+    if (stream.readRawData(buffer, sizeof(buffer)) != sizeof(buffer)) {
+        qWarning() << "Parsing SeratoMarkersEntry failed:"
+                   << "unable to read bytes 10..16";
+        return nullptr;
+    }
+
+    stream >> colorRaw >> type >> isLocked;
+
+    const QRgb color = colorToRgb(colorRaw);
```

Comment:
should we rename the function to colorRawToRgb() as well? Or is there a better name for this interesting format?

Reply:
Maybe `seratoColorToRgb()`?